src/order.c

```diff
@@ -0,0 +1,4148 @@
+  uint8_t* p_bytes = (uint8_t*) p_lazy_bytes->p_data;
+
+  lazy_vec_initialize(p_lazy_counts);
+  R_xlen_t* p_counts = (R_xlen_t*) p_lazy_counts->p_data;
+  memset(p_counts, 0, p_lazy_counts->size);
+
+  void* p_x_chunk = int_maybe_copy(p_x, p_lazy_x_chunk, size, copy);
+  int_adjust(p_x_chunk, decreasing, na_last, size);
+
+  int_radix_order(
+    p_x_chunk,
+    p_o,
+    p_x_aux,
+    p_o_aux,
+    p_bytes,
+    p_counts,
+    p_group_infos,
+    size
+  );
+}
+
+// If we aren't copying, we expect that `p_lazy_x_chunk` already contains
+// the integer values. This happens for character ordering where we store
+// the truelength values in `p_lazy_x_chunk`.
+static inline
+void* int_maybe_copy(const int* p_x,
+                     struct lazy_vec* p_lazy_x_chunk,
+                     R_xlen_t size,
+                     bool copy) {
+  if (!copy) {
+    return p_lazy_x_chunk->p_data;
+  }
+
+  lazy_vec_initialize(p_lazy_x_chunk);
+
+  void* p_x_chunk = p_lazy_x_chunk->p_data;
+
+  memcpy(p_x_chunk, p_x, size * sizeof(int));
+
+  return p_x_chunk;
+}
+
+// -----------------------------------------------------------------------------
+
+static inline uint32_t int_map_to_uint32(int x);
+
+/*
+ * - Shifts the integer elements of `p_x` in a way that correctly maintains
+ *   ordering for `na_last` and `decreasing`
+ *
+ * - After shifting, also maps the elements from `int32_t` to `uint32_t` and
+ *   stores them back in `p_x`.
+ *
+ * - Used before both the integer insertion sort and radix sort, which both
+ *   expect their input to already have been "adjusted" for `na_last` and
+ *   `decreasing` and expect a `uint32_t` pointer input.
+ *
+ * - If `na_last = true`, `NA` is always the maximum element, so we set it to
+ *   `UINT32_MAX`. In that case, we also shift all non-NA values down by 1 to
+ *   make room for it (defined by `na_shift`).
+ *
+ * - If `na_last = false`, we set `NA` to the minimum element of 0.
+ *
+ * - The multiplication by `direction` applies to non-NA values and correctly
+ *   orders inputs based on whether we are in a decreasing order or not.
+ */
+static
+void int_adjust(void* p_x,
+                const bool decreasing,
+                const bool na_last,
+                const R_xlen_t size) {
+  const int direction = decreasing ? -1 : 1;
+  const uint32_t na_u32 = na_last ? UINT32_MAX : 0;
+  const int na_shift = na_last ? -1 : 0;
+
+  const int* p_x_int = (const int*) p_x;
+  uint32_t* p_x_u32 = (uint32_t*) p_x;
+
+  for (R_xlen_t i = 0; i < size; ++i) {
+    int elt = p_x_int[i];
+
+    if (elt == NA_INTEGER) {
+      p_x_u32[i] = na_u32;
+      continue;
+    }
+
+    elt = elt * direction + na_shift;
+
+    p_x_u32[i] = int_map_to_uint32(elt);
+  }
+}
+
+
+#define HEX_UINT32_SIGN_BIT 0x80000000u
+
+// Flipping the sign bit is all we need to do to map in an order preserving way.
+// [INT32_MIN, INT32_MAX] => [0, UINT32_MAX]
+static inline uint32_t int_map_to_uint32(int x) {
+  return x ^ HEX_UINT32_SIGN_BIT;
+}
+
```

Comment:
I used this as the location for switching between int->uint32_t. Other functions like `int_radix_order()` take `uint32_t* p_x` as input. I did this for clarity when thinking about the fact that you shift by 8 bits 4 times = 32 bits total.

If you returned a `unsigned int` here, would you change `int_radix_order()` to also use `unsigned int*` rather than `uint32_t*`? I'd imagine so, and since we already assert that the size of `int` matches our expectations, it should be fine.

Reply:
oh no need for action here. The suggestion is cleaner because it doesn't assume the width of integers, but as you've said we make this assumption in many places now.

> What is unimplemented_int_width?

An identifier that will cause a compiler error on platforms that are not handled by the `#if`.